Fix tests in test/cql-pytest/ that fail on Cassandra

nyh opened the pull request:
As a general rule, tests in test/cql-pytest shouldn't just pass on Scylla - they also should not fail on Cassandra; A test that fails on Cassandra may indicate that the test is wrong, or that Scylla's behavior is wrong and the test just enshrines that wrong behavior. Each time we see a test fail on Cassandra we need to check if this is not the case. We also have special markers scylla_only and cassandra_bug to put on tests that we know _should_ fail on Cassandra because it is missing some Scylla-only feature or there is a bug in Cassandra, respectively. Such tests will be xfailed/skipped when running on Cassandra, and not report failures.

Unfortunately, over time more several tests got into our suite in that did not pass on Cassandra. In this series I went over all of them, and fixed each to pass - or be skipped - on Cassandra, in a way that each patch explains.

Fixes #16027

nyh commented:
I invited several people who wrote some of the tests involved in this PR to review it.

ptrsmrn commented on test/cql-pytest/test_keyspace.py:
```diff
@@ -106,10 +106,21 @@ def test_alter_keyspace_invalid(cql, this_dc):
+# Cassandra 4.1 and above, a missing replication_factor *is* allowed,
+# because there is a default_keyspace_rf configuration. See issue #16028.
+def test_alter_keyspace_missing_rf(cql, this_dc, scylla_only):
+    with new_test_keyspace(cql, "WITH REPLICATION = { 'class' : 'NetworkTopologyStrategy', '" + this_dc + "' : 1 }") as keyspace:
```
I'd suggest to assign RF=3, for the sake of "good usage examples", i.e. having NetworkTopology RF=3 by default.

denesb replied:
This won't work on a single-node cluster, like the one used in `cql-pytest`.

Deexie commented on test/cql-pytest/test_compaction_strategy_validation.py:
```diff
@@ -6,38 +6,48 @@
 # Tests for compaction strategy validation
```
I addressed the points from the commit message under the issue (https://github.com/scylladb/scylladb/issues/16027#issuecomment-1808044240). I would leave it as is. Changing of scylla's `bucket_low` message is worth considering, but that's out of scope for this PR.

nyh replied:
Ok, so I'm leaving my partial patch for this test as I did it. It means that even after my patch, this test will fail on Cassandra. It's not a disaster - nobody runs it on Cassandra anyway (it's not part of our CI), and hopefully another day you'll look into these issues. None of this is urgent at all.

Deexie reviewed (approved):
LGTM (compaction strategy validation part)

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - Build
:white_check_mark: - Sanity Tests
:white_check_mark: - Unit Tests
#### Build Details:
- Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4663/
- Duration: 2 hr 9 min
- Builder: spider1.cloudius-systems.com

denesb commented on test/cql-pytest/test_describe.py:
```diff
@@ -591,6 +619,7 @@ def test_type_quoting(cql):
         assert f"CREATE TYPE \"{ks_name}\".\"{name}\"" in desc
         assert f"CREATE TYPE \"{ks_name}\".{name}" not in desc
         assert f"CREATE TYPE {ks_name}.\"{name}\"" not in desc
+        # This was buggy in Cassandra - see CASSANDRA-19019
```
Looks like this was fixed in a later version.

nyh replied:
Yes, I see now, thanks. I'll fix the comment / marker. Although it's a bit frustrating the test will fail on older Cassandra, we shouldn't be in the business of tracking Cassandra bugs in our test suite, we can just assume it's run on a recent version of Cassandra.

nyh replied:
I upgraded my Cassandra to the latest version of Cassandra 4 (4.1.3), and indeed, it works there. I'm changing the test accordingly.

denesb replied:
Maybe add a note about this to `test/cql-pytest/README.md` (in a follow-up PR). Also, in general, I'm missing a "Running the tests against Cassandra" section in said file anyway.

nyh replied:
I do have a few sentences about Cassandra spread around the README.md, and even in the "philosophy" section a whole part (5) about "Run your test against Cassandra". But maybe I should split that part into a separate section, higher up the file?

I could also explain how to compile Cassandra on modern Linux distros, which isn't trivial because of the Java 17 mess (although you can also take a pre-compiled version of Java and not deal with the compilation).

nyh replied:
I looked again at the text of test/cql-pytest/README.md and saw that almost everything I could possibly say about Cassandra is already written there. Except how to actually compile Cassandra. Could you please tell me a bit more what you were missing there? Or maybe the existing text was just hard for you to find among all the other explanations? Or maybe you were missing the part I admit is missing - how to compile Cassandra?

denesb replied:
When I had a quick look earlier today, I saw a big block of text, and no header standing out, mentioning "Cassandra". Indeed as I read into the text, I do see you talk about it already, so I guess it is fine.

> how to compile Cassandra?

I don't think this is important to describe there, people can find out about that on the C* documentation I suppose.

nyh commented:
Rebased and re-pushed. For one test which discovered a Cassandra bug so was marked `cassandra_bug`, the bug was actually fixed recently (it was fixed in Cassandra 4.1.2, I was running Cassandra 4.1.1...), so I removed the `cassandra_bug` tag.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - Build
:white_check_mark: - Sanity Tests
:white_check_mark: - Unit Tests
#### Build Details:
- Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4701/
- Duration: 2 hr 42 min
- Builder: spider7.cloudius-systems.com
